src/socket/tcp.rs

```diff
@@ -2285,11 +2286,9 @@ impl<'a> Socket<'a> {
                 // to be sent again.
                 self.remote_last_seq = self.local_seq_no;
 
-                // Clear the `should_retransmit` state. If we can't retransmit right
-                // now for whatever reason (like zero window), this avoids an
-                // infinite polling loop where `poll_at` returns `Now` but `dispatch`
-                // can't actually do anything.
-                self.timer.set_for_idle(cx.now(), self.keep_alive);
+                // Set next time of retransmission. `set_for_retransmit` will exponentially increase it.
+                self.timer
+                    .set_for_retransmit(cx.now(), self.rtte.retransmission_timeout());
```

Comment:
i'm not sure if restarting the retransmit timer here is 100% correct, because it's not yet guaranteed we're going to send a packet. Actually transmitting is done further below, which might fail because there's not enough buffers in the device for example. 

i've opened  #1023 which does this a slightly different way: it still does `set_for_idle` here so we only restart the retransmit timer if we did actually send a retransmission packet (because it's done after sending), and keeps the `delay` variable outside the timer (in the RTTE struct) so it doesn't get cleared.